accounts-db/src/accounts_index.rs

```diff
@@ -1149,6 +1149,26 @@ impl<T: IndexValue, U: DiskIndexValue + From<T> + Into<T>> AccountsIndex<T, U> {
         self.get_bin(pubkey).get_internal(pubkey, callback)
     }
 
+    /// Gets the index's entry for `pubkey`, with `ancestors` and `max_root`,
+    /// and applies `callback` to it
+    pub fn get_with_and_then<R>(
+        &self,
+        pubkey: &Pubkey,
+        ancestors: Option<&Ancestors>,
+        max_root: Option<Slot>,
+        should_add_to_in_mem_cache: bool,
+        mut callback: impl FnMut((Slot, T)) -> R,
+    ) -> Option<R> {
+        self.get_and_then(pubkey, |entry| {
+            let index_value = entry.and_then(|entry| {
+                let slot_list = entry.slot_list.read().unwrap();
+                self.latest_slot(ancestors, &slot_list, max_root)
+                    .map(|found_index| callback(slot_list[found_index]))
```

Comment:
in theory, this could copy `(Slot, T)`, release the slot list read lock, and the index read lock, and then call the callback. But, that would be a new locking pattern and could introduce race conditions.